fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -487,40 +519,62 @@ public static long getTo(@Nonnull LucenePartitionInfoProto.LucenePartitionInfo p
     /**
      * Re-balance full partitions, if applicable.
      *
-     * @return void future
+     * @param start The continuation at which to resume rebalancing, as returned from a previous call to
+     * {@code rebalancePartitions}.
+     * @return a tuple of the next group that needs repartitioning
      */
     @Nonnull
-    public CompletableFuture<Void> rebalancePartitions() {
-        if (!isPartitioningEnabled()) {
-            return AsyncUtil.DONE;
-        }
-
+    public CompletableFuture<RecordCursorContinuation> rebalancePartitions(RecordCursorContinuation start) {
         // This function will iterate the grouping keys
-        final ScanProperties scanProperties = ScanProperties.FORWARD_SCAN.with(
-                props -> props.clearState().setReturnedRowLimit(1));
-
-        final Range range = state.indexSubspace.range();
-        final KeyRange keyRange = new KeyRange(range.begin, range.end);
-        final Subspace subspace = state.indexSubspace;
         final KeyExpression rootExpression = state.index.getRootExpression();
 
         if (! (rootExpression instanceof GroupingKeyExpression)) {
-            return processPartitionRebalancing(Tuple.from());
+            return processPartitionRebalancing(Tuple.from()).thenApply(result -> {
+                if (result.getLeft() > 0) {
+                    // we did something, repeat
+                    return RecordCursorStartContinuation.START;
+                } else {
+                    return RecordCursorEndContinuation.END;
+                }
+            });
         }
 
         GroupingKeyExpression expression = (GroupingKeyExpression) rootExpression;
         final int groupingCount = expression.getGroupingCount();
 
+        final ScanProperties scanProperties = ScanProperties.FORWARD_SCAN.with(
+                props -> props.clearState().setReturnedRowLimit(1));
+
+        final Range range = state.indexSubspace.range();
+        final KeyRange keyRange = new KeyRange(range.begin, range.end);
+        final Subspace subspace = state.indexSubspace;
         try (RecordCursor<Tuple> cursor = new ChainedCursor<>(
                 state.context,
                 lastKey -> FDBDirectoryManager.nextTuple(state.context, subspace, keyRange, lastKey, scanProperties, groupingCount),
                 Tuple::pack,
                 Tuple::fromBytes,
-                null,
+                start.toBytes(),
```

Comment:
It would probably be good if this could pass in `start` on it's own, but I didn't want to pull that into this PR.